### PR TITLE
next: Allow write winrate etc. to the comment of the sgf by append-winrate-to-comment

### PR DESCRIPTION
**Pull request by zsalch**
Closes: #308 
Base on: #368 

Support save the winrate etc. into the comment of Sgf (need turn on "append-winrate-to-comment" in the config.txt)
format as followings:

Move \<move number>
\<Winrate> (\<Last Move Rate Difference>)
(\<Weight name> / \<Playouts>)

for example:
W[dd]C[Move 2
53.5% (-0.1%)
(elf_v1 / 18k playouts)

other comments];

### CODE REVIEW

File: src/main/java/featurecat/lizzie/rules/BoardData.java [hunk 1]
```diff
@@ -72,6 +72,9 @@ public static BoardData empty(int size) {
    */
   public void addProperty(String key, String value) {
     SGFParser.addProperty(properties, key, value);
+    if ("N".equals(key) && comment.isEmpty()) {
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/BoardData.java]: Should the line above be moved in an else branch?

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 2]
```diff
@@ -307,6 +308,17 @@ private static void saveToStream(Board board, Writer writer) throws IOException
             "KM[%s]PW[%s]PB[%s]DT[%s]AP[Lizzie: %s]",
             komi, playerW, playerB, date, Lizzie.lizzieVersion));
 
+    // For append the winrate to the comment of sgf, maybe need to update the Winrate
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: *To* append the winrate to the comment of sgf *we might* need to update the Winrate

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 3]
```diff
@@ -307,6 +308,17 @@ private static void saveToStream(Board board, Writer writer) throws IOException
             "KM[%s]PW[%s]PB[%s]DT[%s]AP[Lizzie: %s]",
             komi, playerW, playerB, date, Lizzie.lizzieVersion));
 
+    // For append the winrate to the comment of sgf, maybe need to update the Winrate
+    if (Lizzie.config.appendWinrateToComment) {
+
+      // Update winrate
+      Leelaz.WinrateStats stats = Lizzie.leelaz.getWinrateStats();
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: After this change this code would be duplicated in 5 places, place / nextMove / nextMove (overloaded) / previousMove. Could you please factor this out into an `updateWinrate` method to get ride of all this duplication?

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 4]
```diff
@@ -421,6 +438,62 @@ private static String generateNode(Board board, BoardHistoryNode node) throws IO
...
+      if (Lizzie.config.handicapInsteadOfWinrate) {
+        double currHandicapedWR = Leelaz.winrateToHandicap(100 - curWR);
+        double lastHandicapedWR = Leelaz.winrateToHandicap(lastWR);
+        lastMoveDiff = String.format(": %.2f", currHandicapedWR - lastHandicapedWR);
+      } else {
+        lastMoveDiff = String.format("(%.1f%%)", 100 - lastWR - curWR);
+      }
+    }
+
+    // Format:
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: Duplicate information with the javadoc above

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 5]
```diff
@@ -421,6 +438,62 @@ private static String generateNode(Board board, BoardHistoryNode node) throws IO
...
+   */
+  private static String formatComment(BoardHistoryNode node) {
+    BoardData data = node.getData();
+    String engine = Lizzie.leelaz.currentWeight();
+
+    // Playouts
+    String playouts = Lizzie.frame.getPlayoutsString(data.playouts);
+
+    // Last winrate
+    BoardData lastNode = node.previous().get().getData();
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: How do you know that it's safe to call `.get()` here?

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 6]
```diff
@@ -421,6 +438,62 @@ private static String generateNode(Board board, BoardHistoryNode node) throws IO
...
+    boolean validLastWinrate = (lastNode != null && lastNode.playouts > 0);
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: `lastNode` cannot be `null`.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 7]
```diff
@@ -421,6 +438,62 @@ private static String generateNode(Board board, BoardHistoryNode node) throws IO
...
+      }
+    }
+
+    // Format:
+    // Move <Move number> <Winrate> (<Last Move Rate Difference>)
+    // (<Weight name> / <Playouts>)
+    String wf = "Move %d\n%s %s\n(%s / %s playouts)";
+    String nc = String.format(wf, data.moveNumber, curWinrate, lastMoveDiff, engine, playouts);
+
+    if (data.comment != null) {
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/SGFParser.java]: `comment` cannot be null, you probably want `!data.comment.isEmpty()` here.

File: src/main/java/featurecat/lizzie/rules/BoardData.java (hunk 1 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: No, I want to keep the N attribute.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 2 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Fixed

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 3 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Yes, fixed.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 4 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Removed

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 6 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 7 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java (hunk 5 above)
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: I see. But still, it's dangerous to has assumptions across methods like this without documentation. If you really don't won't to handle the case where a node doesn't have a previous node, the I would add it to the javadoc and start the method witha a `if (!node.previous().isPresent()) throw new IllegalArgumentException("...")`.

Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Ok, fixed.

**Comment by featurecat** [comment]
I'm merging, because we need to hurry things along for the 0.6 release.

**Pull request merged**